### ISSUE TITLE
Question: How can I create ArrayField of MoneyField? 

### ISSUE THREAD

**Issue by rajeshr188** [django-money/django-money#599]
I tried it as per django dics but it saves only numeric [] of decimals in postgres without currency column

**Comment by Stranger6667** [comment]
This use case is not supported. `django-money` implicitly adds `CurrencyField` on the same table. With arrays, it is not clear whether it is expected to store amount/currency pairs (1) or share the same currency for all those values (2). 

The first option probably won't work directly as in Postgres, arrays are homogeneous, and storing things in separate columns will require synchronization, which may not be trivial (but surely possible). 
Alternatively, we could implement `Money` as a custom type for Postgres (somewhat like in [my talk](https://www.youtube.com/watch?v=g12fkl9NCrM&ab_channel=DjangoConEurope) at DjangoConEurope), then it will be possible to put them into an array. Still, we'll need to maintain shims for databases that don't support user-defined types (e.g., MySQL). It is hard to estimate this approach in terms of complexity, but it surely will increase it quite a bit. However, personally, I like the idea of having a self-contained `Money` type on the DB level (the PG built-in money type is a different thing). Some parts of the current implementation will be much simpler (e.g., we could use the Lookup API instead of manually converting keyword arguments).

Then, the second option requires some work; indeed, I didn't check how it currently behaves under the hood in this case but based on your comment in #597, I assume that the money field is just not visible to the logic that decides whether `CurrencyField` should be added or not.

I hope that it clarifies the status quo.

**Comment by rajeshr188** [comment]
Yes thanks for clarifying. 
But what was the reason behind not creating composite type on db side in django-money?

One reason I understand is from your talk is to make django-money database agnostic since mysql doesn't support composite types. 
If my app will only have postgresdb as backend and I create a custom modelfield derived from moneyed.Money that creates composite type in db, what corner-cases I should take care of? 
In other words if you were to implement this project only for postgres backend how the implementation would have changed?

**Comment by rajeshr188** [comment]
I have implemented composite type money_value in postgres and corresponding MoneyvalueField in django that saves Money object in db as money_value(amount, currency) type and we can retreive back Money object from db. 

And an ArrayField of MoneyvalueField that saves array of Money objects in db and we can retreive back Moneyobject[(amount, currency),(amount, currency)], it's working for timebeing. Any suggestions for improvement s?

**Comment by Stranger6667** [comment]
> But what was the reason behind not creating composite type on db side in django-money?

Mostly historical ones, the project is quite old and at some point, there was no Lookup API and other things and a custom type didn't pop up anywhere (I didn't know about it at all until ~2017).

> If my app will only have postgresdb as backend and I create a custom modelfield derived from moneyed.Money that creates composite type in db, what corner-cases I should take care of?

Depends on the use case, but straightforwardly implementing the current django-money lookups via the Lookup API should be sufficient. However, there are always corner cases that are hard to predict :)

> In other words if you were to implement this project only for postgres backend how the implementation would have changed?
I'd use a custom PG type for it, I even had an unpublished prototype somewhere :) As far as I remember it was much simpler & less hacky at least on the lookup side of things with a little performance penalty. I will look for that project :) But generally speaking, a lot of code that synchronises NUMERIC & VARCHAR(3) (it could be CHAR(3) btw, as we work only with 3 symbol currencies ) would be gone.

> I have implemented composite type money_value in postgres and corresponding MoneyvalueField in django that saves Money object in db as money_value(amount, currency) type and we can retreive back Money object from db.

Awesome!

> Any suggestions for improvement s?

I'd be happy to see the code first, before discussing improvements :)

**Comment by rajeshr188** [comment]
[here](https://gist.github.com/rajeshr188/9cf4143fa1113166a1b5ed3a5c859483) is the barebone implementation gist of what i implemented.i am able to insert Money[] in to db as composite money_value(amount,currency) and get array of Money objects out of db in to a list of Money.pls i dont know what is lookups im reading through django docs and django-money.
